clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -1007,16 +1047,33 @@ public Map<TopicPartition, OffsetAndMetadata> fetchCommittedOffsets(final Set<To
             if (pendingCommittedOffsetRequest != null) {
                 future = pendingCommittedOffsetRequest.response;
             } else {
-                future = sendOffsetFetchRequest(partitions);
-                pendingCommittedOffsetRequest = new PendingCommittedOffsetRequest(partitions, generationForOffsetRequest, future);
+                future = sendOffsetFetchRequest(nonCachedPartitions);
+                pendingCommittedOffsetRequest = new PendingCommittedOffsetRequest(nonCachedPartitions, generationForOffsetRequest, future);
             }
             client.poll(future, timer);
 
             if (future.isDone()) {
                 pendingCommittedOffsetRequest = null;
 
                 if (future.succeeded()) {
-                    return future.value();
+                    Map<TopicPartition, OffsetAndMetadata> freshOffsets = future.value();
+
+                    // update cache for assigned partitions that are not cached yet
+                    for (TopicPartition nonCachedAssignedPartition: nonCachedAssignedPartitions) {
+                        if (!this.subscriptions.isAssigned(nonCachedAssignedPartition)) {
+                            // it is possible that the topic is no longer assigned when the response is received,
+                            // in this case we do not update the cache with the fresh value
+                            continue;
+                        }
+
+                        OffsetAndMetadata offset = freshOffsets.get(nonCachedAssignedPartition);
+                        if (offset != null) { // it is possible that the offset and metadata were not fetched
+                            this.partitionOffsetsCache.put(nonCachedAssignedPartition, offset);
```

Comment:
Some high level comment here.
You added offset commit cache when fetching offsets. Currently, we will fetch offset commit when consumer proactively fetch it (won't happen frequently), and also when the consumer first time got assigned a partition, so it needs to know where to start to fetch. And after consumer starts, it'll periodically commit offsets. That means, if we cache this data, it'll be out-dated soon after consumer committed offsets.

On the other hand, if we cache at the place where we committed offsets (i.e. `OffsetCommitResponseHandler`), we should be able to get the latest committed offsets for each partition whenever the consumer commits it. Does that make sense?

Reply:
Thank for you for this, I get the gist of what you are saying and it makes sense. I'll go have a look at this again and re-visit it with that new understanding or I'll come back with additional questions if I need of more understanding.

Reply:
Hi @showuon I forgot to mention or call your attention on this line https://github.com/apache/kafka/blob/401fb417bf60864e6d380f979d268d895c5ad727/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L1422 which already updates the cache. Is that what you've in mind?